test/extensions/filters/network/thrift_proxy/filters/payload_to_metadata/payload_to_metadata_filter_test.cc

```diff
@@ -0,0 +1,1226 @@
+  EXPECT_TRUE(!rhs.empty());
+  for (auto const& entry : rhs) {
+    EXPECT_NE(obj.fields().find(entry.first), obj.fields().end());
+    EXPECT_EQ(obj.fields().at(entry.first).string_value(), entry.second);
+  }
+  return true;
+}
+
+MATCHER_P(MapEqNum, rhs, "") {
+  const ProtobufWkt::Struct& obj = arg;
+  EXPECT_TRUE(!rhs.empty());
+  for (auto const& entry : rhs) {
+    EXPECT_NE(obj.fields().find(entry.first), obj.fields().end());
+    EXPECT_EQ(obj.fields().at(entry.first).number_value(), entry.second);
+  }
+  return true;
+}
+
+} // namespace
+
+using namespace Envoy::Extensions::NetworkFilters;
+
+class PayloadToMetadataTest : public testing::Test {
+public:
+  void initializeFilter(const std::string& yaml, bool expect_type_inquiry = true) {
+    envoy::extensions::filters::network::thrift_proxy::filters::payload_to_metadata::v3::
+        PayloadToMetadata proto_config;
+    TestUtility::loadFromYaml(yaml, proto_config);
+    const auto& filter_config = std::make_shared<Config>(proto_config);
+    filter_ = std::make_shared<PayloadToMetadataFilter>(filter_config);
+    filter_->setDecoderFilterCallbacks(decoder_callbacks_);
+    if (expect_type_inquiry) {
+      EXPECT_CALL(decoder_callbacks_, downstreamTransportType()).WillOnce(Return(transport_));
+      EXPECT_CALL(decoder_callbacks_, downstreamProtocolType()).WillOnce(Return(protocol_));
+    } else {
+      EXPECT_CALL(decoder_callbacks_, downstreamTransportType()).Times(0);
+      EXPECT_CALL(decoder_callbacks_, downstreamProtocolType()).Times(0);
+    }
+  }
+
+  // Request payload
+  // {
+  //   first_field: 1,
+  //   second_field: "two" or string_value,
+  //   third_field: {
+  //     f1: true,
+  //     f2: 2,
+  //     ...
+  //     f6: 6,
+  //     f7: "seven",
+  //     f8: map,
+  //     f9: list,
+  //     f10: set
+  //   }
+  // }
+  void writeMessage(MessageMetadata& metadata, Buffer::Instance& buffer,
+                    std::string string_value = "two") {
+    Buffer::OwnedImpl msg;
+    ProtocolPtr proto = NamedProtocolConfigFactory::getFactory(protocol_).createProtocol();
+    metadata.setProtocol(protocol_);
+    metadata.setMethodName("foo");
+    metadata.setMessageType(MessageType::Call);
+    metadata.setSequenceId(0);
+
+    proto->writeMessageBegin(msg, metadata);
+    proto->writeStructBegin(msg, "wrapper");
+    proto->writeFieldBegin(msg, "first_field", FieldType::I64, 1);
+    proto->writeInt64(msg, 1);
+    proto->writeFieldEnd(msg);
+
+    proto->writeFieldBegin(msg, "second_field", FieldType::String, 2);
+    proto->writeString(msg, string_value);
+    proto->writeFieldEnd(msg);
+
+    proto->writeFieldBegin(msg, "third_field", FieldType::Struct, 3);
+
+    proto->writeStructBegin(msg, "payload");
+    proto->writeFieldBegin(msg, "f1", FieldType::Bool, 1);
+    proto->writeBool(msg, true);
+    proto->writeFieldEnd(msg);
+
+    proto->writeFieldBegin(msg, "f2", FieldType::Byte, 2);
+    proto->writeByte(msg, 2);
+    proto->writeFieldEnd(msg);
```

Comment:
nit: i think we have helpers for making this stuff shorter, iirc there's something like:

```
addField(msg, "f2", FieldType::Byte, 2);
```

that will reduce the boilerplate... if not, would be nice to have!

Reply:
Unfortunately we don't have a helper to the best of my knowledge.
The definition is here
https://github.com/envoyproxy/envoy/blob/9e4ddc314b78ecbad21645e0128e7e0a05964926/source/extensions/filters/network/thrift_proxy/protocol.h#L278

We used this pattern here and there.
Here's an example
https://github.com/envoyproxy/envoy/blob/9e4ddc314b78ecbad21645e0128e7e0a05964926/test/extensions/filters/network/thrift_proxy/conn_manager_test.cc#L263

I do agree with you that we want some helper but that needs some careful deliberation so we won't use the helper wrong. 
We want same experience among struct, container(map, list, set) and single fields so we can do something like
``` cpp
ProtocolHelper helper(proto, msg, "wrapper_struct_name"); // start the wrapper
helper.addInt64Field("first_field", 1, 1);
helper.addStringField("second_field", 2, "two");
helper.addStructField("third_field", 3, [&]() {
   helper.addBoolField("f1", 1, true);
   // ...
});
helper.addMapField("fourth_field", 4, FieldType::I32, {{1, 1}, {2,2});
helper.done(); //finish the wrapper

transport->encodeFrame(buffer, metadata, msg);
```

There we can fully hide `writeFieldBegin` otherwise it's so easy to mis-use the helper.
I would foreseen it takes some time on the lambda callback, testing and template programming on the container field.
If you agree, I can do this as a follow up and refactor the existing tests here and there.